### PR TITLE
DBZ-2092 Pub/Sub support in server

### PR DESCRIPTION
**Pull request by jpechane**
https://issues.redhat.com/browse/DBZ-2092

### CODE REVIEW

File: debezium-server/src/test/java/io/debezium/server/PubSubIT.java [hunk 1]
```diff
@@ -0,0 +1,123 @@
...
+import com.google.pubsub.v1.PushConfig;
+import com.google.pubsub.v1.TopicName;
+
+import io.debezium.server.events.ConnectorCompletedEvent;
+import io.debezium.server.events.ConnectorStartedEvent;
+import io.debezium.util.Testing;
+import io.quarkus.test.junit.QuarkusTest;
+
+/**
+ * Integration test that verifies basic reading from PostgreSQL database and writing to Kinesis stream.
```
Comment (gunnarmorling) [inline debezium-server/src/test/java/io/debezium/server/PubSubIT.java:38]:
```suggestion
 * Integration test that verifies basic reading from PostgreSQL database and writing to a Google Cloud PubSub stream.
```

File: documentation/modules/ROOT/pages/operations/debezium-server.adoc [hunk 2]
```diff
@@ -255,7 +255,7 @@ The configuration will need to contain the list of transformations, implementati
 === Sink configuration
 
 Sink configuration is specific for each sink type.
-Currently the only supported sink is https://aws.amazon.com/kinesis/[Amazon Kinesis].
+Currently the only supported sinks are https://aws.amazon.com/kinesis/[Amazon Kinesis] and https://cloud.google.com/pubsub/[Google Pub/Sub].
```
Comment (gunnarmorling) [inline documentation/modules/ROOT/pages/operations/debezium-server.adoc:258]:
```suggestion
Currently the only supported sinks are https://aws.amazon.com/kinesis/[Amazon Kinesis] and https://cloud.google.com/pubsub/[Google Cloud Pub/Sub].
```

File: documentation/modules/ROOT/pages/operations/debezium-server.adoc [hunk 3]
```diff
@@ -305,9 +306,64 @@ When the alternative implementations are not available then the default ones are
...
+|[[kinesis-ext-stream-name-mapper]]<<kinesis-ext-stream-name-mapper, `io.debezium.server.StreamNameMapper`>>
 |
-|Custom implementation maps the planned destination (topic) name into a physical Kinesis stream name. By default the same name is used.
+|Custom implementation maps the planned destination (topic) name into a physical Kinesis stream name.
+By default the same name is used.
+
+|=======================
+
+
+==== Google Pub/Sub
```
Comment (gunnarmorling) [inline documentation/modules/ROOT/pages/operations/debezium-server.adoc:317]:
```suggestion
==== Google Cloud Pub/Sub
```

File: documentation/modules/ROOT/pages/operations/debezium-server.adoc [hunk 4]
```diff
@@ -305,9 +306,64 @@ When the alternative implementations are not available then the default ones are
...
+
+Google Pub/Sub is an implementation of messaging/eventing system designed for scalable batch and stream processing applications.
```
Comment (gunnarmorling) [inline documentation/modules/ROOT/pages/operations/debezium-server.adoc:319]:
```suggestion
Google Cloud Pub/Sub is an implementation of messaging/eventing system designed for scalable batch and stream processing applications.
```

File: documentation/modules/ROOT/pages/operations/debezium-server.adoc [hunk 5]
```diff
@@ -305,9 +306,64 @@ When the alternative implementations are not available then the default ones are
...
+
+[cols="35%a,10%a,55%a",options="header"]
+|=======================
+|Interface
+|CDI classifier
+|Description
+
+|[[pubsub-pub-builder]]<<pubsub-pub-builder, `io.debezium.server.pubsub.PubSubChangeConsumer.PublisherBuilder`>>
+|`@CustomConsumerBuilder`
+|A calss that provides custom configured instance of a `Publisher` used to send messages to a dedicated topic.
```
Comment (gunnarmorling) [inline documentation/modules/ROOT/pages/operations/debezium-server.adoc:362]:
```suggestion
|A class that provides custom configured instance of a `Publisher` used to send messages to a dedicated topic.
```

File: debezium-server/src/test/java/io/debezium/server/TestConfigSource.java [hunk 6]
```diff
@@ -21,24 +21,39 @@
...
-        e2eTest.put("debezium.source.database.hostname", TestDatabase.POSTGRES_HOST);
-        e2eTest.put("debezium.source.database.port", Integer.toString(TestDatabase.POSTGRES_PORT));
-        e2eTest.put("debezium.source.database.user", TestDatabase.POSTGRES_USER);
-        e2eTest.put("debezium.source.database.password", TestDatabase.POSTGRES_PASSWORD);
-        e2eTest.put("debezium.source.database.dbname", TestDatabase.POSTGRES_DBNAME);
-        e2eTest.put("debezium.source.database.server.name", "testc");
-        e2eTest.put("debezium.source.schema.whitelist", "inventory");
-        e2eTest.put("debezium.source.table.whitelist", "inventory.customers");
+        pubsubTest.put("debezium.sink.type", "pubsub");
+        pubsubTest.put("debezium.sink.", KINESIS_REGION);
```
Comment (gunnarmorling) [inline debezium-server/src/test/java/io/debezium/server/TestConfigSource.java:31]: Seems odd to use KINESIS_REGION here?

File: documentation/modules/ROOT/pages/operations/debezium-server.adoc [hunk 7]
```diff
@@ -305,9 +306,64 @@ When the alternative implementations are not available then the default ones are
...
+
+|[[pubsub-type]]<<pubsub-type, `debezium.sink.type`>>
+|
+|Must be set to `pubsub`.
+
+|[[pubsub-project-id]]<<pubsub-projectid, `debezium.sink.pubsub.projectid`>>
+|_system-wide default project id_
+|A project name in which the target topics are created.
+
+|[[pubsub-ordering]]<<pubsub-ordering, `debezium.sink.pubsub.ordering.enabled`>>
```
Comment (gunnarmorling) [inline documentation/modules/ROOT/pages/operations/debezium-server.adoc:336]: Re ordering, I found that one: https://cloud.google.com/pubsub/docs/ordering. Should we link it here? I haven't full understood it yet, could you perhaps give a summary in layman's terms?

Comment (jpechane) [reply documentation/modules/ROOT/pages/operations/debezium-server.adoc:336]: https://googleapis.dev/java/google-cloud-clients/latest/index.html?com/google/cloud/pubsub/v1/Publisher.html

Comment (gunnarmorling) [reply documentation/modules/ROOT/pages/operations/debezium-server.adoc:336]: Worth adding a link to that to the docs, I think.

**Comment by gunnarmorling** [comment]
@jpechane, do you plan to follow up on the comments re module split and README for testing as part of this PR?

**Comment by jpechane** [comment]
@gunnarmorling I am working on it now

**Comment by jpechane** [comment]
@gunnarmorling Done, as always the law of unintended conseqeunces has hit and it was necessary to solve impact on testing and distribution packing.

**Comment by gunnarmorling** [comment]
@jpechane, any idea re the test failure?

**Comment by jpechane** [comment]
@gunnarmorling It is very rare intermittent failure, unrelated to the changes. Not the first time I see it.

**Comment by gunnarmorling** [comment]
Ok, logged https://issues.redhat.com/browse/DBZ-2126 for the test failure.

File: debezium-server/debezium-server-pubsub/pom.xml [hunk 8]
```diff
@@ -0,0 +1,129 @@
...
+
+        <!-- Target systems -->
+        <dependency>
+            <groupId>com.google.cloud</groupId>
+            <artifactId>google-cloud-pubsub</artifactId>
+        </dependency>
+        <dependency>
+            <groupId>com.google.cloud</groupId>
+            <artifactId>google-cloud-core</artifactId>
+        </dependency>
```
Comment (gunnarmorling) [inline debezium-server/debezium-server-pubsub/pom.xml:32]: What are the licenses? Do they allow for redistribution?

File: debezium-server/debezium-server-core/pom.xml [hunk 9]
```diff
@@ -0,0 +1,123 @@
...
+<project xmlns="http://maven.apache.org/POM/4.0.0" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd">
+    <parent>
+        <groupId>io.debezium</groupId>
+        <artifactId>debezium-server</artifactId>
+        <version>1.2.0-SNAPSHOT</version>
+        <relativePath>../pom.xml</relativePath>
+    </parent>
+    <modelVersion>4.0.0</modelVersion>
+    <artifactId>debezium-server-core</artifactId>
+    <name>Debezium Standalone Quarkus Server Core</name>
```
Comment (gunnarmorling) [inline debezium-server/debezium-server-core/pom.xml:11]:
```suggestion
    <name>Debezium Server Core</name>
```

File: debezium-server/debezium-server-kinesis/pom.xml [hunk 10]
```diff
@@ -0,0 +1,115 @@
...
+<project xmlns="http://maven.apache.org/POM/4.0.0" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd">
+    <parent>
+        <groupId>io.debezium</groupId>
+        <artifactId>debezium-server</artifactId>
+        <version>1.2.0-SNAPSHOT</version>
+        <relativePath>../pom.xml</relativePath>
+    </parent>
+    <modelVersion>4.0.0</modelVersion>
+    <artifactId>debezium-server-kinesis</artifactId>
+    <name>Debezium Standalone Quarkus Server Amazon Kinesis Sender</name>
```
Comment (gunnarmorling) [inline debezium-server/debezium-server-kinesis/pom.xml:11]:
```suggestion
    <name>Debezium Server Amazon Kinesis Sink Adapter</name>
```

File: debezium-server/debezium-server-pubsub/pom.xml [hunk 11]
```diff
@@ -0,0 +1,129 @@
...
+<project xmlns="http://maven.apache.org/POM/4.0.0" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd">
+    <parent>
+        <groupId>io.debezium</groupId>
+        <artifactId>debezium-server</artifactId>
+        <version>1.2.0-SNAPSHOT</version>
+        <relativePath>../pom.xml</relativePath>
+    </parent>
+    <modelVersion>4.0.0</modelVersion>
+    <artifactId>debezium-server-pubsub</artifactId>
+    <name>Debezium Standalone Quarkus Server Google Cloud Pub/Sub Sender</name>
```
Comment (gunnarmorling) [inline debezium-server/debezium-server-pubsub/pom.xml:11]:
```suggestion
    <name>Debezium Server Google Cloud Pub/Sub Sink Adapter</name>
```

File: debezium-server/debezium-server-pubsub/pom.xml (hunk 8 above)
Comment (jpechane) [reply debezium-server/debezium-server-pubsub/pom.xml:32]: ASL 2 and BSD

Comment (gunnarmorling) [reply debezium-server/debezium-server-pubsub/pom.xml:32]: Ok, thx.

**Comment by gunnarmorling** [comment]
Thx!
